src/core/autodiff.cc

```diff
@@ -0,0 +1,387 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tc/core/autodiff.h"
+#include "tc/core/tc2halide.h"
+#include "tc/lang/parser.h"
+#include "tc/lang/sema.h"
+#include "tc/lang/tc_format.h"
+#include "tc/lang/tree_views.h"
+
+#include <unordered_set>
+#include <vector>
+
+namespace tc {
+
+using namespace lang;
+
+static const lang::SourceRange dummyRange{std::make_shared<std::string>(""),
+                                          0,
+                                          0};
+
+int32_t getTcType(Halide::Type t) {
+  if (t.is_int()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_INT64;
+      case 32:
+        return TK_INT32;
+      case 16:
+        return TK_INT16;
+      case 8:
+        return TK_INT8;
+    }
+  } else if (t.is_uint()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_UINT64;
+      case 32:
+        return TK_UINT32;
+      case 16:
+        return TK_UINT16;
+      case 8:
+        return TK_UINT8;
+    }
+  } else if (t.is_float()) {
+    switch (t.bits()) {
+      case 64:
+        return TK_DOUBLE;
+      case 32:
+        return TK_FLOAT;
+    }
+  }
+  throw std::runtime_error("Unknown Halide type");
+}
+
+void findAccessedTensors(
+    std::unordered_set<std::string>& read_only,
+    const TreeRef& tree) {
+  if (tree->kind() == TK_ACCESS) {
+    read_only.insert(Access(tree).name().name());
+  } else {
+    for (const TreeRef& subtree : tree->trees()) {
+      findAccessedTensors(read_only, subtree);
+    }
+  }
+}
+
+void assertNoWriteAfterRead(Def def) {
+  std::unordered_set<std::string> read_only;
+  // Inputs are always read-only
+  for (Param input : def.params())
+    read_only.insert(input.ident().name());
+  for (Comprehension comp : def.statements()) {
+    findAccessedTensors(read_only, comp.rhs());
+    auto lhs_name = comp.ident().name();
+    if (read_only.count(lhs_name) > 0)
+      throw std::runtime_error(
+          "AD not supported in TCs that write to a value after reading it");
```

Comment:
So this would not work for FCRelu?
```
def fcrelu(float(N, M) I, float(M, P) W, float(P) Bias) -> (O) {
  O(n, p) +=! W(r_m, p) * I(n, r_m)
  O(n, p)  = O(n, p) + B(p)
}
```
Is this a fundamental limitation or a future TODO?